apps/aechannel/src/aesc_fsm.erl

```diff
@@ -1790,8 +1795,53 @@ adjust_ttl(undefined) ->
 adjust_ttl(TTL) when is_integer(TTL), TTL >= 0 ->
     TTL.
 
+-spec curr_hash_and_height() -> {aec_blocks:block_header_hash(),
+                                 aec_blocks:height()}.
+curr_hash_and_height() ->
+    TopHeader = aec_chain:top_header(),
+    {ok, Hash} = aec_headers:hash_header(TopHeader),
+    Height = aec_headers:height(TopHeader), 
+    {Hash, Height}.
+
+-spec curr_height() -> aec_blocks:height().
 curr_height() ->
-    aec_headers:height(aec_chain:top_header()).
+    {_, Height} = curr_hash_and_height(),
+    Height.
+
+pick_hash(#data{block_hash_delta = #bh_delta{not_newer_than = NNT}}) ->
+    TopHeader = aec_chain:top_header(),
+    Height = aec_headers:height(TopHeader), 
+    %% use upper limit
+    {ok, Header} = aec_chain:get_key_header_by_height(max(Height - NNT, 0)),
+    {ok, Hash} = aec_headers:hash_header(Header),
+    Hash.
+
+load_pinned_env(?NOT_SET_BLOCK_HASH) ->
+    {_OnChainEnv, _OnChainTrees} =
+        aetx_env:tx_env_and_trees_from_top(aetx_contract);
+load_pinned_env(BlockHash) ->
+    try aetx_env:tx_env_and_trees_from_hash(aetx_contract, BlockHash) of
+        {_OnChainEnv, _OnChainTrees} = Pinned ->
+            Pinned
+    catch error:{badmatch,error} -> erlang:error(unknown_block_hash)
+    end.
+
+-spec pick_pinned_env(map(), #data{}) ->
+    {aec_blocks:block_header_hash(), aetx_env:env(), aec_trees:trees()}. 
+pick_pinned_env(#{block_hash := ?NOT_SET_BLOCK_HASH}, _D) ->
```

Comment:
The semantics seem misleading because the function can also return a non-pinned environment and respective info.